### DISCUSSION TITLE
HOW TO GENERATE THE STOICHIOMETRIC MATRIX FOR STYRENE POLYMERIZATION IN FREE RADICAL.

### DISCUSSION THREAD

**Discussion by edwardiwuchukwu** [opencobra/cobrapy#1140] [Q&A]
Good day.
Referring to the above heading, I have been working on generating the stoichiometric matrix using the COBRAPY python module. The propagation, chain transfer and termination reaction(the combination is more important) are growing chains so I will really appreciate your assistance in providing the code or pseudocode for the reaction below:

![free_radical_polymerization_2](https://user-images.githubusercontent.com/85592612/145108931-bf7ded71-c97c-43a9-b503-5ec68af7d344.PNG)

I look forward to your timely response.
Kind regards,
Edward.

**Comment by matthiaskoenig** [comment]
Hi Edward,

you should have a look how to add species and reactions to a model
https://cobrapy.readthedocs.io/en/latest/building_model.html#Model,-Reactions-and-Metabolites
Here you would write a loop over n and m and create the respective species and reactions.

I am not sure a constraint-based model is what you are interested in. This looks much more like a kinetic model, especially because you have rates/parameters for the transitions `k_d`, k_i`, `k_p`, ....
Because you work with polymers a rule-based approach to the model generation could be what you are looking for.

So in summary:
- you should make sure you want a constraint-based approach
- you have to dynamically generate your species and reactions in a loop

Best Matthias

**Comment by edwardiwuchukwu** [comment]
Hello Matthias,
Thank you for the swift response to the posted question.
Yes, this is a kinetic-based model. You were right with the problem description. 
Can you suggest an alternative approach to producing the stoichiometric matrix to kinetic models of the type?

Kind regards,
Edward

**Comment by matthiaskoenig** [comment]
Hi @teddiclax ,
you could use sbmlutils (disclaimer: developed by me) to easily create the SBML.
https://github.com/matthiaskoenig/sbmlutils 

You basically write a python script which creates the SBML objects, i.e. species, reactions, parameters, ...
For an example of a linear chain see the following
https://github.com/matthiaskoenig/sbmlutils/blob/develop/src/sbmlutils/examples/linear_chain.py
You would do something similar but iterate in a double loop over n and m and add the respective objects.
This is NOT a rule based model, i.e. you would just create the objects up to the n you would need in your model (at some point your ki will all be zero, if you have a stable system without running against infinite polymer length, so you don't have to model larger polymers).

For an introductory tutorial see:
https://www.youtube.com/watch?v=aGtXCK2KtX8

For rule based approaches see for instance pySB, but I am not sure these can deal with infinity polymers either.
Let me know if you need help with sbmlutils to create your model.

Best Matthias

**Comment by edwardiwuchukwu** [comment]
Good day Mattias,
Thank you very for the response. I will take a look at this and revert back
to you.

Kind regards,
Edward.
